### PR TITLE
Support IP Based Endpoints and Changes serviceVersion Methods

### LINKED ISSUE #5823: Support IP-Style URLs (enables Azurite)
Azurite uses a special URL format that we have to support. More investigation is needed but details will be posted here.

### PR DESCRIPTION
**Pull request by alzimmermsft**
Fixes #5823 and #5906 

Adds support for IP based endpoints in Blobs and Queues. Changes the `getServiceVersion` methods to return the enum value.

### CODE REVIEW

File: sdk/storage/azure-storage-queue/src/main/java/com/azure/storage/queue/QueueClientBuilder.java [hunk 1]
```diff
@@ -156,7 +129,28 @@ public QueueClient buildClient() {
      * or {@link #sasToken(String) SAS token} has been set.
      */
     public QueueAsyncClient buildAsyncClient() {
-        return new QueueAsyncClient(constructImpl(), queueName, accountName);
+        Objects.requireNonNull(queueName, "'queueName' cannot be null.");
```
Comment (conniey) [inline sdk/storage/azure-storage-queue/src/main/java/com/azure/storage/queue/QueueClientBuilder.java]: iirc, you have that Utility.assertNotNull

File: sdk/storage/azure-storage-queue/src/main/java/com/azure/storage/queue/BuilderHelper.java [hunk 2]
```diff
@@ -76,6 +82,61 @@ static void configureConnectionString(String connectionString, Consumer<String>
         credentialSetter.accept(new StorageSharedKeyCredential(accountName, accountKey));
     }
 
+    static void parseEndpoint(String endpoint, Consumer<String> endpointSetter, Consumer<String> accountNameSetter,
```
Comment (conniey) [inline sdk/storage/azure-storage-queue/src/main/java/com/azure/storage/queue/BuilderHelper.java]: I'd prefer to return an object with all these parts populated, similar to your bloburl parts. THis is like having multiple "out String foo" method parameters

File: sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedFlux.java [hunk 3]
```diff
@@ -47,7 +48,7 @@
 
     public static final int DATA_COUNT = 40;
 
-    public static final int DOWNLOAD_SIZE = 64; // The total size of the "download" after expanding the requested range.
+    private static final int DOWNLOAD_SIZE = 64; // The total size of the "download" after expanding the requested range.
```
Comment (srnagar) [inline sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedFlux.java]: Adding units in comments or in the constant name would make is easy to understand what 64 here represents. Bytes, KB  etc.
`DOWNLOAD_SIZE_IN_BYTES` or `DOWNLOAD_SIZE_IN_KB` etc.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobUrlParts.java [hunk 4]
```diff
@@ -17,13 +17,19 @@
...
 import java.util.TreeMap;
+import java.util.regex.Pattern;
 
 /**
  * This class represents the components that make up an Azure Storage Container/Blob URL. You may parse an
  * existing URL into its parts with the {@link #parse(URL)} class. You may construct a URL from parts by calling {@link
  * #toURL()}.
  */
 public final class BlobUrlParts {
+    private static final Pattern IP_URL_PATTERN = Pattern
```
Comment (srnagar) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobUrlParts.java]: nit: `IP_V4_URL_PATTERN`

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobUrlParts.java [hunk 5]
```diff
@@ -243,25 +249,33 @@ public URL toURL() throws MalformedURLException {
...
+     * @throws IllegalArgumentException If {@code url} is a malformed {@link URL}.
      */
-    public static BlobUrlParts parse(String url, ClientLogger logger) {
+    public static BlobUrlParts parse(String url) {
         try {
             return parse(new URL(url));
         } catch (MalformedURLException e) {
-            throw logger.logExceptionAsError(new IllegalArgumentException("Please double check the URL format. URL: "
-                + url));
+            throw new IllegalArgumentException("Please double check the URL format. URL: " + url);
```
Comment (srnagar) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobUrlParts.java]: Error message should be `"Invalid URL format. URL: " + url`

File: sdk/storage/azure-storage-queue/src/main/java/com/azure/storage/queue/BuilderHelper.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-queue/src/main/java/com/azure/storage/queue/BuilderHelper.java]: Done

File: sdk/storage/azure-storage-queue/src/main/java/com/azure/storage/queue/QueueClientBuilder.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-queue/src/main/java/com/azure/storage/queue/QueueClientBuilder.java]: Done

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobUrlParts.java (hunk 4 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobUrlParts.java]: Done

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobUrlParts.java (hunk 5 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobUrlParts.java]: Done

**Pull request merged**